### ISSUE TITLE
[WAITING ON CHECKLIST] [$1000] Web - Secondary contact method transition bug - clicking on a contact method that has yet to be validated opens the magic code input (validation) page with a sped up animation

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#23593]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to settings > profile > contact method
2. Click on new contact method
3. Enter an email address
4. Click add
5. Click on the email address you just entered
6. Observe how the transition animation to the magic code input (validation) page is sped up/doesn’t work properly. If you look closely you can see that the text (sometimes) wobbles

## Expected Result:
The page transition should be animated like other page transitions, without it being sped up.

## Actual Result:
The page transition appears to not be animated/is sped up.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome 
- [ ] MacOS / Desktop

**Version Number:** 1.3.44.0
**Reproducible in staging?:** n/a
**Reproducible in production?:** n/a
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/d508a806-8032-4764-b653-18391f9abfbd

**Expensify/Expensify Issue URL:**
**Issue reported by:** @joh42
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1690229176544109

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b86e6c6c9fa20453</li>
        <li>Upwork Job ID: 1684009715954917376</li>
        2023-07-26<li>Automatic offers: </li>
<ul>
<li>joh42 | Contributor | 25814019</li>
<li>joh42 | Reporter | 25814020</li>
</ul></li>
    </ul>
</details>

**Comment by kbecciv** [comment]
**Proposal**: @joh42
**Slack conversation**: https://expensify.slack.com/archives/C049HHMV9SM/p1690229176544109

# Proposal
### Please re-state the problem that we are trying to solve in this issue.

We are trying to solve the sped up animation when the secondary contact method validation page is opened.

### What is the root cause of that problem?

The root cause of the problem is that focus is given to the magic code text input instantly, before the transition animation has finished.

### What changes do you think we should make in order to solve the problem?

The solution requiring the least amount of changes to the codebase would be to always pass `shouldDelayFocus={true}` from the [BaseValidateCodeForm](https://github.com/Expensify/App/blob/main/src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.js) component to the [MagicCodeInput](https://github.com/Expensify/App/blob/main/src/components/MagicCodeInput.js) component. Currently that is only passed as true for Android. Alternatively we could pass it as true for only Android and Chrome, but that would be more complex.
https://github.com/Expensify/App/blob/beebb67612cea0201f4fb1b6b17840cb5893d754/src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.js#L142

That would be enough to solve this issue. However, while we're at it, we might want to improve the code a bit:

We can simplify the [MagicCodeInput](https://github.com/Expensify/App/blob/main/src/components/MagicCodeInput.js) component by removing the following `useEffect`

https://github.com/Expensify/App/blob/2c15e9758c081563eb755156680342f1cfbb74e4/src/components/MagicCodeInput.js#L156-L176

Instead we could pass `shouldDelayFocus` along to the first [TextInput](https://github.com/Expensify/App/tree/a656c82d0ebf1d3954575084555a004efc47035b/src/components/TextInput) component, as opposed to not passing `autoFocus` if `shouldDelayFocus` is true:
https://github.com/Expensify/App/blob/2c15e9758c081563eb755156680342f1cfbb74e4/src/components/MagicCodeInput.js#L327-L329

The [TextInput](https://github.com/Expensify/App/tree/a656c82d0ebf1d3954575084555a004efc47035b/src/components/TextInput) component [already supports](https://github.com/Expensify/App/blob/a656c82d0ebf1d3954575084555a004efc47035b/src/components/TextInput/BaseTextInput.js#L67-L69) the `shouldDelayFocus` prop, so removing that redundant functionality from [MagicCodeInput](https://github.com/Expensify/App/blob/main/src/components/MagicCodeInput.js) would make the code a bit DRYer.
https://github.com/Expensify/App/blob/a656c82d0ebf1d3954575084555a004efc47035b/src/components/TextInput/BaseTextInput.js#L67-L70

### What alternative solutions did you explore? (Optional)

There is currently an ongoing [Slack discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1689957344255689) regarding a general approach to solving this issue. The two alternate solutions to knowing when we should set focus is to use either `InteractionManager.runAfterInteractions` or `onTransitionEnd`. Both of those would require more changes than using the already implemented `shouldDelayFocus`. I nonetheless think we should wait for that discussion to conclude, and then consider if we should go with the solution that is deemed optimal even if it is more complex to implement.

**Comment by jliexpensify** [comment]
Tested on newest version and yes, it is more sped up compared to the default contact method transition.

**Comment by joh42** [comment]
Thanks for reposting my proposal from Slack! I'm just going to append the following to the alternative solutions section:
At present the discussion in Slack is leaning towards standardizing focusing using the [ScreenWrapper](https://github.com/Expensify/App/blob/main/src/components/ScreenWrapper/index.js) component's `onEntryTransitionEnd` prop. If we do decide that it should be implemented for this issue, I think it should be done in the [TextInput](https://github.com/Expensify/App/tree/a656c82d0ebf1d3954575084555a004efc47035b/src/components/TextInput) component to make sure the code is as DRY as possible. The tradeoff to keep in mind is that if we keep using `shouldDelayFocus`, we would not have to change that component at all.

**Comment by jliexpensify** [comment]
Bumo @parasharrajat got some proposals!

**Comment by joh42** [comment]
So that issue is more related to the alternative solution I mentioned rather than the animation issue

**Comment by joh42** [comment]
@parasharrajat I'm a bit confused, what part of that proposal was not covered by my proposal already? According to the contributing guidelines:
```Note: Before submitting a proposal on an issue, be sure to read any other existing proposals. ALL NEW PROPOSALS MUST BE DIFFERENT FROM EXISTING PROPOSALS. The difference should be important, meaningful or considerable.```

I clearly mentioned using `transitionEnd`, and while I agree that the edit I made there is incorrect, per https://expensify.slack.com/archives/C01GTK53T8Q/p1690295235023339?thread_ts=1690291230.607589&cid=C01GTK53T8Q proposals should not get into the nitty gritty, but rather identify the root cause and propose a solution, which I thought I did by mentioning `onTransitionEnd` in my original proposal.

**Comment by joh42** [comment]
Also, I'm new here, so I'm genuinely curious: should proposals contain high-level solutions, or should they actually delve into what I would consider the nitty gritty, implementation details?

**Comment by joh42** [comment]
To give further clarification here, in my comment I explicitly wrote:
> Thanks for reposting my proposal from Slack! I'm just going to append the following to the alternative solutions section: At present the discussion in Slack is leaning towards standardizing focusing using the [ScreenWrapper](https://github.com/Expensify/App/blob/main/src/components/ScreenWrapper/index.js) component's `onEntryTransitionEnd` prop.

After that, I posited that we could perhaps use `didScreenTransitionEnd`. This was in no way meant to remove the alternate solution of using `onEntryTransitionEnd`.

**Comment by joh42** [comment]
Reading through this, I realize it's quite hard to follow my multiple comments. As such, I'm going to summarize everything below.

I fail to see how explaining how to focus using `onEntryTransitionEnd` fulfills **ALL NEW PROPOSALS MUST BE DIFFERENT FROM EXISTING PROPOSALS. The difference should be important, meaningful or considerable.** Focusing using `onEntryTransitionEnd` has already been implemented 7 times already by different contributors, I did not think I needed to explain how to do it once more, but that mentioning it on a high level would be enough. Here are the existing implementations:
https://github.com/Expensify/App/blob/975146fa30bc65a19df8e0cca525af32ce2ceda5/src/pages/EditRequestDescriptionPage.js#L30
https://github.com/Expensify/App/blob/975146fa30bc65a19df8e0cca525af32ce2ceda5/src/pages/tasks/NewTaskDetailsPage.js#L76
https://github.com/Expensify/App/blob/975146fa30bc65a19df8e0cca525af32ce2ceda5/src/pages/settings/Payments/AddPayPalMePage.js#L68
https://github.com/Expensify/App/blob/975146fa30bc65a19df8e0cca525af32ce2ceda5/src/pages/iou/MoneyRequestDescriptionPage.js#L96
https://github.com/Expensify/App/blob/975146fa30bc65a19df8e0cca525af32ce2ceda5/src/pages/settings/Report/RoomNamePage.js#L82
https://github.com/Expensify/App/blob/975146fa30bc65a19df8e0cca525af32ce2ceda5/src/pages/settings/Payments/AddDebitCardPage.js#L102
https://github.com/Expensify/App/blob/975146fa30bc65a19df8e0cca525af32ce2ceda5/src/pages/tasks/TaskTitlePage.js#L66

Rather, it seems like that contributor is just getting into the [nitty gritty](https://expensify.slack.com/archives/C01GTK53T8Q/p1690295235023339?thread_ts=1690291230.607589&cid=C01GTK53T8Q)

It seems like the C+ member might have missed that I wrote the line below:
> At present the discussion in Slack is leaning towards standardizing focusing using the [ScreenWrapper](https://github.com/Expensify/App/blob/main/src/components/ScreenWrapper/index.js) component's onEntryTransitionEnd prop.

And instead only saw another (incorrect) alternate solution that I posited. I might be mistaken, but I thought that since it's worded 'alternative solution**S**' in the proposal template, you could post more than one, and if one is incorrect that would not discard the others.

**Comment by thienlnam** [comment]
Thank you both for the proposals and discussion for this issue - these types of situations are always tricky, and unfortunately there isn't a great way to handle this. 

Tienifr has handled this situation before in previous issues and has had a proven method that has worked before. 

However, joh42 mentioned it briefly in the original proposal. Since you're a newer contributor here, I'd recommend for future issues, delving deeper into one or two solutions that you know will work and then the C+ can be more confident on recommending your proposal.

I'm going to go with @joh42 for this issue because I think you had the general idea correct, but are still figuring out the process and working on refining your future proposals.

I hope that's alright with everyone cc @tienifr / @parasharrajat

**Comment by joh42** [comment]
Thanks for the guidance @parasharrajat / @thienlnam , I appreciate it. Going forward I’ll zero-in on one or two solutions per proposal, and make sure I clearly state what I mean with each one. As for this issue, you can expect a PR to be ready for review tomorrow.

**Comment by joh42** [comment]
Regarding a compensation split, I would like to mention the following:

Unless I’m mistaken, the so-called ‘exact solution’ is incorrect. The SignInPage does not and should not use `shouldDelayFocus` as there is no transition animation. Rather, it should be focused as soon as it mounts. Default focusing using `autoFocus` is not impacted by this issue.

Besides, relying on `onEntryTransitionEnd` to delay focusing on the SigninPage would not work, as the BaseValidateCodeForm component, which contains the MagicCodeInput, is conditionally rendered, which means that `onEntryTransitionEnd` is not fired after that component is shown. If that page required delayed focusing, we would have to check `didScreenTransitionEnd` when the MagicCodeInput component mounts.

https://github.com/Expensify/App/blob/8cdb506e97e8382f4bc55b0f48784f576fbb2789/src/pages/signin/SignInPage.js#L153

To be clear, this issue is only related to `shouldDelayFocus`, and the only change required to fix the issue at hand for the affected pages is removing `shouldDelayFocus` in favor of focusing using `onEntryTransitionEnd`.

As such, the ‘exact solution’ is incorrect as it involves a page that does not have this issue, is not affected by the solution to this issue, and said solution would not even work for that page.

It’s also worth noting that this is not the first time you submit a proposal that is similar to an existing one and then ask for a compensation split. In issue [https://github.com/Expensify/App/issues/16582](https://github.com/Expensify/App/issues/16582#issuecomment-1511690827) your request for a compensation split was rebuffed by an Expensify employee with the following motivation:

[Quote from comment 1](https://github.com/Expensify/App/issues/16582#issuecomment-1511640631):

>The extra details in your proposal were good, but IMO both proposals were suggesting the same solution. So a 50/50 split doesn't seem that fair to me, given that your proposal followed an existing one and you didn't have to complete the checklist.

[Quote from comment 2](https://github.com/Expensify/App/issues/16582#issuecomment-1511690827):

>Typically I might go with a more detailed proposal, but I don't want to encourage contributors to 'swipe' existing proposals just by adding more details, or to add the full diff to the prosal [sic] when it isn't necessary.

That being said, if @thienlnam disagrees with me here and thinks a compensation split is appropriate I’ll go along with it, but I think there is a case to be made against it.

**Comment by thienlnam** [comment]
There are plenty of opportunities for more work here, and I've already made my decision. Let's work on other bugs rather than continuing to go back and fourth on this issue which wastes everyone's time. 

We'll continue on in this issue as planned - thanks!

**Comment by joh42** [comment]
I just noticed that the Storybook stories for the MagicCodeInput is broken on main. Should I write that in Slack? Or should I report that as a bug? @parasharrajat / @thienlnam 
<img width="1436" alt="Screenshot 2023-08-01 at 19 35 26" src="https://github.com/Expensify/App/assets/138504087/ec445270-b48c-4591-a898-2914aa1c0459">

**Comment by joh42** [comment]
Checking off the Storybook checkbox even though I cannot actually test it, thanks for responding in Slack

**Comment by jliexpensify** [comment]
Ok, here's a payment summary:

- Bug Reporter / Contributor: @joh42 $250 + $1000
- C+: @parasharrajat $1000 - request on New.Dot

[Upwork job](https://www.upwork.com/jobs/~01b86e6c6c9fa20453)

**Comment by jliexpensify** [comment]
@joh42 can you add your full name (Johannes Idarsson) to your GH profile as well? Cheers!

**Comment by joh42** [comment]
Added, thanks @jliexpensify

**Comment by jliexpensify** [comment]
Hmm, looks like it was just deployed so we'd wait another 7 days to pay this one

**Comment by jliexpensify** [comment]
Hmm it was actually deployed to production last week...odd. @parasharrajat let me know if there's a regression test you want GH'd.

**Comment by jliexpensify** [comment]
Paid Johannes, job closed. Just awaiting checklist.

**Comment by jliexpensify** [comment]
Cheers, done!

**Comment by JmillsExpensify** [comment]
Reviewed the details for @parasharrajat. This $1,000 payment is approved for NewDot based on the BZ summary above.
